File: src/web/src/render/effect/ShockwaveEffect.test.ts
import { describe, it, expect } from 'vitest';
import { ShockwaveEffect } from './ShockwaveEffect.js';

const WW = ShockwaveEffect.WAVE_WIDTH;

function makePixels(w: number, h: number, fill = 0): Uint8ClampedArray {
  const arr = new Uint8ClampedArray(w * h * 4);
  if (fill !== 0) arr.fill(fill);
  return arr;
}

function copyPixels(src: Uint8ClampedArray): Uint8ClampedArray {
  return new Uint8ClampedArray(src);
}

describe('ShockwaveEffect', () => {
  it('isActive is false before trigger', () => {
    const e = new ShockwaveEffect(() => {});
    expect(e.isActive()).toBe(false);
  });

  it('isActive is true immediately after trigger', () => {
    const e = new ShockwaveEffect(() => {});
    e.trigger(50, 50);
    expect(e.isActive()).toBe(true);
  });

  it('pixels far from the ring are unchanged', () => {
    const w = 300, h = 300;
    const src = makePixels(w, h, 0xAB);
    const dst = copyPixels(src);

    const e = new ShockwaveEffect(() => {});
    e.trigger(150, 150);
    // radius starts at WAVE_WIDTH (~20), so pixels at corners are untouched.
    e.applyToPixels(src, dst, w, h);

    expect(dst[0]).toBe(src[0]);
    expect(dst[(h - 1) * w * 4 + (w - 1) * 4]).toBe(src[(h - 1) * w * 4 + (w - 1) * 4]);
  });

  it('centre pixel (d=0) is not displaced', () => {
    const w = 300, h = 300;
    const src = new Uint8ClampedArray(w * h * 4);
    for (let i = 0; i < src.length; i++) src[i] = (i % 251) as number;
    const dst = copyPixels(src);

    const e = new ShockwaveEffect(() => {});
    e.trigger(150, 150);
    e.applyToPixels(src, dst, w, h);

    const centreBase = (150 * w + 150) * 4;
    expect(dst[centreBase]).toBe(src[centreBase]);
  });

  it('pixels on the ring are displaced (gradient src)', () => {
    const w = 400, h = 400;
    const cx = 200, cy = 200;
    const src = new Uint8ClampedArray(w * h * 4);
    for (let y = 0; y < h; y++) {
      for (let x = 0; x < w; x++) {
        const base = (y * w + x) * 4;
        src[base]     = x & 0xFF;
        src[base + 3] = 255;
      }
    }
    const dst = copyPixels(src);

    const e = new ShockwaveEffect(() => {});
    e.trigger(cx, cy);

    // Place a test pixel exactly on the ring: distance = radius = WAVE_WIDTH, so distToRing=0.
    // WAVE_WIDTH is the initial radius, so pixels at distance ~WAVE_WIDTH from centre are on the ring.
    e.applyToPixels(src, dst, w, h);

    // At (cx + WW, cy): d ≈ WW, distToRing ≈ 0, bell ≈ 1 → should be displaced.
    const testBase = (cy * w + (cx + Math.round(WW))) * 4;
    const origR  = src[testBase];
    const displR = dst[testBase];
    expect(displR).not.toBe(origR);
  });

  it('stop deactivates the effect', () => {
    const e = new ShockwaveEffect(() => {});
    e.trigger(50, 50);
    expect(e.isActive()).toBe(true);
    e.stop();
    expect(e.isActive()).toBe(false);
  });
});

File: src/test/java/com/ullgren/modern/simple3d/render/effect/ShockwaveEffectTest.java
package com.ullgren.modern.simple3d.render.effect;

import org.junit.jupiter.api.Test;
import static org.junit.jupiter.api.Assertions.*;

/**
 * Unit tests for {@link ShockwaveEffect}.
 * <p>
 * Tests operate directly on {@code int[]} pixel arrays — no AWT or Renderer involved.
 */
public class ShockwaveEffectTest {

  @Test
  public void isActive_beforeTrigger_isFalse() {
    assertFalse(new ShockwaveEffect(() -> {}).isActive());
  }

  @Test
  public void isActive_afterTrigger_isTrue() {
    ShockwaveEffect effect = new ShockwaveEffect(() -> {});
    effect.trigger(50, 50);
    assertTrue(effect.isActive(), "Should be active immediately after trigger");
  }

  @Test
  public void applyToPixels_atRingRadius_pixelDisplaced() {
    // At trigger time the ring is at radius = WAVE_WIDTH.
    // Pixels within one belt-width of the ring should be displaced.
    int w = 200, h = 200;
    int[] src = new int[w * h];
    for (int y = 0; y < h; y++)
      for (int x = 0; x < w; x++)
        src[y * w + x] = 0xFF000000 | (x & 0xFF);
    int[] dst = new int[w * h];
    System.arraycopy(src, 0, dst, 0, src.length);

    ShockwaveEffect effect = new ShockwaveEffect(() -> {});
    effect.trigger(100, 100);
    effect.applyToPixels(src, dst, w, h);

    // At trigger time the ring sits at d = WAVE_WIDTH from centre.
    // The pixel exactly at the ring centre has bell = 1 → maximum displacement.
    int ringX = Math.round(100 + ShockwaveEffect.WAVE_WIDTH);
    assertNotEquals(src[100 * w + ringX], dst[100 * w + ringX],
        "Pixel at the ring centre should be displaced");
  }

  @Test
  public void applyToPixels_farFromRing_pixelsUnchanged() {
    int w = 400, h = 400;
    int[] src = new int[w * h];
    int[] dst = new int[w * h];
    for (int i = 0; i < src.length; i++) src[i] = 0xFF0000FF;
    System.arraycopy(src, 0, dst, 0, src.length);

    ShockwaveEffect effect = new ShockwaveEffect(() -> {});
    effect.trigger(200, 200);
    effect.applyToPixels(src, dst, w, h);

    // Corners are far from any ring radius — must be unchanged
    assertEquals(src[0],                         dst[0],                         "Top-left corner must be unchanged");
    assertEquals(src[(h - 1) * w + (w - 1)],     dst[(h - 1) * w + (w - 1)],    "Bottom-right corner must be unchanged");
  }

  @Test
  public void applyToPixels_centrePixel_notDisplaced() {
    int w = 200, h = 200;
    int[] src = new int[w * h];
    for (int i = 0; i < src.length; i++) src[i] = 0xFF000000 | i;
    int[] dst = new int[w * h];
    System.arraycopy(src, 0, dst, 0, src.length);

    ShockwaveEffect effect = new ShockwaveEffect(() -> {});
    effect.trigger(100, 100);
    effect.applyToPixels(src, dst, w, h);

    // d=0 is skipped; the ring starts at WAVE_WIDTH, so the centre is always outside the belt
    assertEquals(src[100 * w + 100], dst[100 * w + 100],
        "Centre pixel must not be displaced");
  }
}
